[Feature] Introduce WooCommerce Wear

ThomazFB opened the pull request:
Summary
==========
This is the feature branch for the [WooCommerce Wear app project](https://github.com/orgs/woocommerce/projects/233/views/1).

It was built through the following PRs:

### Woo PRs

- https://github.com/woocommerce/woocommerce-android/pull/11366
- https://github.com/woocommerce/woocommerce-android/pull/11384
- https://github.com/woocommerce/woocommerce-android/pull/11396
- https://github.com/woocommerce/woocommerce-android/pull/11415
- https://github.com/woocommerce/woocommerce-android/pull/11423
- https://github.com/woocommerce/woocommerce-android/pull/11451
- https://github.com/woocommerce/woocommerce-android/pull/11458
- https://github.com/woocommerce/woocommerce-android/pull/11347
- https://github.com/woocommerce/woocommerce-android/pull/11439
- https://github.com/woocommerce/woocommerce-android/pull/11471
- https://github.com/woocommerce/woocommerce-android/pull/11464
- https://github.com/woocommerce/woocommerce-android/pull/11505
- https://github.com/woocommerce/woocommerce-android/pull/11528
- https://github.com/woocommerce/woocommerce-android/pull/11561
- https://github.com/woocommerce/woocommerce-android/pull/11564
- https://github.com/woocommerce/woocommerce-android/pull/11579
- https://github.com/woocommerce/woocommerce-android/pull/11581
- https://github.com/woocommerce/woocommerce-android/pull/11745
- https://github.com/woocommerce/woocommerce-android/pull/11760

### FluxC and Tooling PRs

- https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/3004
- https://github.com/woocommerce/woocommerce-android/pull/11567
- https://github.com/woocommerce/woocommerce-android/pull/11636
- 
- https://github.com/woocommerce/woocommerce-android/pull/11726

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ThomazFB committed:
Create ParseOrderData use case

ThomazFB committed:
Remove OrderItem usage from ViewModel and use the Use case one

ThomazFB committed:
Configure OrderDetailsViewModel order parsing and handling

ThomazFB committed:
Adjust OrderDetailsScreen to listen to Order data changes

ThomazFB committed:
Fix lint issues

ThomazFB committed:
Fix navigation crashes

ThomazFB committed:
Improve Navigation args declaration

ThomazFB committed:
Add preview and initial style adjustments to OrderDetailsScreen

ThomazFB committed:
Fix and improve Order details data fetching

ThomazFB committed:
Fix lint issues

ThomazFB committed:
Add initial header structure for OrderDetailsScreen

ThomazFB committed:
Add TimeText and text alignments

ThomazFB committed:
Centralize all format operations to the ParseOrderData

ThomazFB committed:
Add Text style to OrderDetailsScreen

ThomazFB committed:
Rename ParseOrderData.kt to FormatOrderData

ThomazFB committed:
Add Order Details string resources

ThomazFB committed:
Fix lint issues

ThomazFB committed:
Add Order Refunds fetching support

ThomazFB committed:
Fix lint issues

ThomazFB committed:
Define WearOrderProduct type

ThomazFB committed:
Introduce GetWearableOrderProducts operation

ThomazFB committed:
Add Order Products support to Wearable Paths

ThomazFB committed:
Handle Order products request from WearOS

ThomazFB committed:
Add the Order ID looping back to the OrderProducts request

ThomazFB committed:
Receive and store Order Products data from phone

ThomazFB committed:
Refactor Order Products requests from WearableConnectionRepository

ThomazFB committed:
Introduce FlowUtils.kt

ThomazFB committed:
Refactor flow with timeout operations

ThomazFB committed:
Rename FlowUtils to FlowExt.kt

ThomazFB committed:
Fix lint issues

ThomazFB committed:
Introduce FetchOrderProducts with data store observation from OrdersRepository

ThomazFB committed:
Improve FetchOrderProducts with dedicated Request class

ThomazFB committed:
Refactor FetchOrders to follow FetchOrderProducts strategy

ThomazFB committed:
Overhaul Store stats request structure

ThomazFB committed:
Improve Stats fetching even further

ThomazFB committed:
Enforce site cohesion inside StatsRepository

ThomazFB committed:
Add missing fixes for the Order Products request

ThomazFB committed:
Merge pull request #11505 from woocommerce/issue/android-wear-order-details-ui

[Watch App] Introduce Order Details UI

ThomazFB committed:
Integrate OrderDetailsViewModel with FetchOrderProducts use case

ThomazFB committed:
Add Products mapping support into FormatOrderData

ThomazFB committed:
Combine site and products data before fetching the Order data

ThomazFB committed:
Configure the UI basics for the Product list

ThomazFB committed:
Adjust combineWithTimeout to allow custom timeout values

ThomazFB committed:
Add Site ID parameter to every request sent to the watch

ThomazFB committed:
Update OrdersRepository to demand for the siteID during the data retrieval

ThomazFB committed:
Adjust Order fetching use cases to also observe the site ID

ThomazFB committed:
Refactor OrderDetailsViewModel flow structures

ThomazFB committed:
Immediately abort the FetchOrderProducts when phone is not available

ThomazFB committed:
Enforce site ID from DataMap for every data received

ThomazFB committed:
Fix stats loading not appearing when site takes too long to be fetched

ThomazFB committed:
Migrate to the wearableStore usage

ThomazFB committed:
Rename ObserveLoginRequest.kt to FetchSiteData.kt

ThomazFB committed:
Fix FetchSiteDataTest

ThomazFB committed:
Improve Order fetch strategy with proper database usage

ThomazFB committed:
Simplify LoginViewModel site fetching strategy

ThomazFB committed:
FIx lint issues

ThomazFB committed:
Start Orders loading immediately after the ViewModel starts

ThomazFB committed:
Add bottom padding to the OrderDetailsScreen

ThomazFB committed:
Add wearableStore usage to the WearableConnectionRepository

ThomazFB committed:
Add remaining string resources from Order Details

ThomazFB committed:
Remove all string literals from OrderDetailsScreen and introduce pluralizedProductsText

ThomazFB committed:
Fix lint issues

ThomazFB committed:
Merge branch 'trunk' into feature/introduce-wear-app

ThomazFB committed:
Update FluxC

ThomazFB committed:
Merge branch 'trunk' into issue/adjust-wear-order-details-ui-data

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  0 Messages
  1 Markdown
-->

## Danger has errored
[!] Invalid `Dangerfile` file: git `--git-dir=/var/lib/buildkite-agent/builds/ci-cluster-linter-i-06f369e55a6217c59-2/automattic/woocommerce-android/.git` `--work-tree=/var/lib/buildkite-agent/builds/ci-cluster-linter-i-06f369e55a6217c59-2/automattic/woocommerce-android` `-c` `core.quotePath=true` `-c` `color.ui=false` `cat-file` `-t` ``  2\>&1
status: pid 3478141 exit 128
output: "fatal: Not a valid object name \n"
```
 #  from Dangerfile:37
 #  -------------------------------------------
 #  
 >  pr_size_checker.check_diff_size(
 #    max_size: 300,
 #  -------------------------------------------
```

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

ThomazFB committed:
Merge branch 'feature/introduce-wear-app' into issue/adjust-wear-order-details-ui-data

ThomazFB committed:
Fix LoginViewModelTest

ThomazFB committed:
Introduce initial structure for FetchOrdersTest

ThomazFB committed:
Improve FetchOrders use case to avoid duplicated updates

ThomazFB committed:
Fix lint issues

ThomazFB committed:
Merge pull request #11528 from woocommerce/issue/adjust-wear-order-details-ui-data

Integrate OrderDetailsViewModel with FetchOrderProducts use case

ThomazFB committed:
Move the ScopedViewModel to the Wear module

ThomazFB committed:
Remove the CoroutineDispatchers.kt usage

ThomazFB committed:
Move Date and Number extensions from commons to the Wear module

ThomazFB committed:
Move the SavedStateFlow code to the FlowExt.kt

ThomazFB committed:
Remove MultiLiveEvent.kt from commons lib

ThomazFB committed:
Finish final steps of commons lib simplification

ThomazFB committed:
Remove complication and tiles code

ThomazFB committed:
Remove commons dependencies

ThomazFB committed:
Abandon direct usage of the OrderEntity inside the Wear module

ThomazFB committed:
Refactor OrdersRepository Entity remapping

ThomazFB committed:
Merge pull request #11561 from woocommerce/issue/simplify-commons-lib

[Watch App] Simplify commons library

ThomazFB committed:
Make every Wear ViewModel a default lifecycle observer

ThomazFB committed:
Configure the main three wear screens to contain the lifecycle observer registration

ThomazFB committed:
Add onResume handlers for Order Details, list and Stats screens

ThomazFB committed:
Add return to avoid requesting something when it's already loading

ThomazFB committed:
Fix selection area of the Order list

ThomazFB committed:
Improve Order Details products list feedback

ThomazFB committed:
Update error strings from Order List and Details

ThomazFB committed:
Add ErrorScreen to Order Details and Orders List

ThomazFB committed:
Introduce the Order list scrollbar

ThomazFB committed:
Fix lint issues

wzieba committed:
Update Wear app versioning

- rely on `version.properties` values for mobile app
- add `w` suffix to `versionName`
- add constant to `versionCode`

wzieba committed:
Add `app` parameter to `build_bundle` lane

wzieba committed:
Merge pull request #11565 from woocommerce/wear_app_versioning

Update Wear app versioning

ThomazFB committed:
Correctly align the localSiteId into the Order shared models

ThomazFB committed:
Completely remove Data Store usage from the Orders list

wzieba committed:
Prefix track name if it's app other than mobile

ThomazFB committed:
Fix Product item count

wzieba committed:
Use constant when referencing Gradle project/directory

ThomazFB committed:
Fix lint issues

wzieba committed:
Remove WOOCOMMERCE prefix from constant

It's shorter now and we're in context of WooCommerce anyway

wzieba committed:
Replace reference by symbol, in favor of string

I'd prefer to not introduce a few ways of doing same thing (app identification), hence let's maybe use String for this task

wzieba committed:
Release beta and production of wear app

ThomazFB committed:
Remove unused string resources

ThomazFB committed:
Add proper error handling to Stats screen

ThomazFB committed:
Fix lint issues

ThomazFB committed:
Merge pull request #11564 from woocommerce/issue/improve-data-sync-handling-in-wear

[Watch App] Introduce UI improvements

ThomazFB committed:
Merge branch 'trunk' into feature/introduce-wear-app

ThomazFB committed:
Update FluxC version

ThomazFB committed:
Add data client to the WearableClientModule

ThomazFB committed:
Introduce AnalyticsTracker initial structure

ThomazFB committed:
Add dataClient support to the PhoneConnectionRepository

ThomazFB committed:
Create AnalyticsTracker initial track function

ThomazFB committed:
Remove unused log

ThomazFB committed:
Detect Analytics Data event coming from the Wear

ThomazFB committed:
Add the Watch Analytics events

ThomazFB committed:
Add Wear Analytics mapping to Phone Analytics structure

AliSoftware committed:
Revert back to having each lane only build one app

Make `build_and_upload_release` and `build_and_upload_beta` take the `app:` as parameter and only build that specific app. That way we will be able to build them independently in CI, as well as only build and upload a wearOS release when we need one (which might not be the case for _every_ version of the mobile app)

AliSoftware committed:
Make `trigger_release_build` accept `app:` param

And also rename `release-builds.yml` pipeline to `release-build.yml` (singular)

Then:
 - Pass that param as `APP_TO_BUILD` env var to the triggered pipeline
 - Make use of that `$APP_TO_BUILD` var in the `release-build.yml` pipeline (`label` and `command`) so it passes the value to the `release-build.sh` script it calls
 - Make the `release-build.sh` script pass `app:"$1"` when it calls the `build_and_upload_google_play` lane
 - Make the `build_and_upload_google_play` lane propagate that `app:` parameter to the `build_and_upload_{beta,release}` lanes it calls

AliSoftware committed:
Add `app:` to `trigger_release_build` call sites

ThomazFB committed:
Add Retry button to the ErrorScreen component

ThomazFB committed:
Add retry support to the OrderDetailsScreen

ThomazFB committed:
Add retry support to the OrdersListScreen

ThomazFB committed:
Improve ErrorScreen with retry button

ThomazFB committed:
Rearrange OrdersListScreen main title to only display when there's orders available

ThomazFB committed:
Small refactor to FetchOrders

ThomazFB committed:
Add dedicated error message filling the whole screen into the StoreStatsScreen

ThomazFB committed:
Wire reload behavior from ViewModel with the StoreStatsScreen

ThomazFB committed:
Improve error handling feedback in StoreStatsScreen

ThomazFB committed:
Refactor the ViewModel basic structure for the Wear app

ThomazFB committed:
Update the LoginScreen strings

AliSoftware committed:
Improve YARD doc of some lanes

AliSoftware committed:
Make Wear app job happen in same CI build (opt-in)

Instead of the WEAR_APP being build by a separate `release-build.yml` CI build with a different `app:` parameter, we now have the `release-builds.yml` pipeline contain 2 jobs, one for the `MOBILE_APP` and one for the `WEAR_APP`, the latter being conditional on the `INCLUDE_WEAR_APP` env var being defined

ThomazFB committed:
Add dedicated paddings for round and not round devices

AliSoftware committed:
Add the app being build in the CI logs

ThomazFB committed:
Add Lightning icon to the LoginScreen

ThomazFB committed:
Adjust error buttons colors

AliSoftware committed:
Fix rubocop violations

ThomazFB committed:
Rename presentation package to compose

ThomazFB committed:
Remove mystore package

ThomazFB committed:
Update Login Screen with correct strings and colors

ThomazFB committed:
Reorganize the LoginScreen vertical placement

ThomazFB committed:
Add Woo bubble with white letters

ThomazFB committed:
Refactor some UI elements

ThomazFB committed:
Adjust the LoginScreen loading

ThomazFB committed:
Adjust OrderDetailsScreen header section

ThomazFB committed:
Adjust OrderDetailsScreen products list

ThomazFB committed:
Final tweaks to the OrderDetailsScreen

ThomazFB committed:
Additional adjustments to text styles

ThomazFB committed:
Fix missing loading during store switch

ThomazFB committed:
Fix lint issues

ThomazFB committed:
Merge pull request #11579 from woocommerce/issue/refactor-error-and-details-UI

[Watch App] Improve error feedback and details UI

ThomazFB committed:
Merge branch 'feature/introduce-wear-app' into issue/add-watch-app-tracks

ThomazFB committed:
Wire Analytics event in the phone side

ThomazFB committed:
Refactor Analytics event mapping

ThomazFB committed:
Add OrdersList tracks events

ThomazFB committed:
Add Order Details tracks events

ThomazFB committed:
Add Login tracks events

ThomazFB committed:
Remaining adjustments for Tracks events

ThomazFB committed:
Fix lint issues

ThomazFB committed:
Refactor Wear AnalyticsEvent names

ThomazFB committed:
Use stored order refunds as the Watch data source

ThomazFB committed:
Fix lint issues

ThomazFB committed:
Reorganize WearAnalyticsEvent event disposition

ThomazFB committed:
Introduce data source tracking events

ThomazFB committed:
Reduce padding in Order Products list

ThomazFB committed:
Fix lint issues

ThomazFB committed:
Merge pull request #11581 from woocommerce/issue/add-watch-app-tracks

[Watch App] Introduce tracks support

ThomazFB committed:
Merge branch 'trunk' into feature/introduce-wear-app

ThomazFB committed:
Update FluxC version

ThomazFB committed:
Adjust Order Details to fetch the products during the Orders presentation

ThomazFB committed:
Introduce Order address data

ThomazFB committed:
Introduce OrderItemAddress empty handling

ThomazFB committed:
Introduce the Customer Address UI to the OrderDetailsScreen

ThomazFB committed:
Fix Order retry screen flashing

ThomazFB committed:
Improve Order Details loading feedback

ThomazFB committed:
Improve spacing between sections in OrderDetailsScreen

ThomazFB committed:
Fix Watch unit tests

ThomazFB committed:
Fix StoreStatsViewModelTest package

ThomazFB committed:
Merge branch 'trunk' into feature/introduce-wear-app

ThomazFB committed:
Fix lint issues

AliSoftware committed:
Merge branch 'feature/introduce-wear-app' into support_wear_app_fastlane

AliSoftware committed:
Merge pull request #11567 from woocommerce/support_wear_app_fastlane

Support WooCommerce-Wear in `fastlane` & Buildkite configuration

AliSoftware committed:
Update release-toolkit to 11.x

AliSoftware committed:
Fix `build_bundle` to build the expected app

We forgot to update the gradle tasks being called to include `:#{app}:` in front of those (`:WooCommerce:assemble` / `:WooCommerce-Wear:assemble` instead of just `assemble`) to tell it to just build said module.

AliSoftware committed:
Add Prototype Build for Wear app

AliSoftware committed:
Fix rubocop violation

ThomazFB committed:
Merge branch 'trunk' into feature/introduce-wear-app

AliSoftware committed:
Fix call to `prototype-build.sh` in pipeline

 - `chmod +x` all the `.buildkite/commands/*` scripts (especially the `prototype-build.sh` one which didn't have the executable bit)
 - Fix the usage of quotes in the `command:` field of the pipeline

ThomazFB committed:
Introduce FetchStatsTest

ThomazFB committed:
Fix lint issues

AliSoftware committed:
Fix flavor option handling in `build_bundle`

Some parts of that lane's implementation (especially `output_aab_file` variable) had `vanilla` flavor hardcoded instead of reading it from the `options[:flavor]` like the rest of the lane's implementation.

Also made that `flavor` option default to `Vanilla` if none was passed explicitly.

ThomazFB committed:
Introduce FetchOrderProductsTest

AliSoftware committed:
More useful metadata in Prototype PR Comment

To differentiate Mobile vs Wear prototype builds/comments

ThomazFB committed:
Introduce Jalapeno variant to Wear module

AliSoftware committed:
Merge branch 'feature/introduce-wear-app' into tooling/wear-app-prototype-build

ThomazFB committed:
Add Jalapeno variant icons and colors

wzieba committed:
Support JaCoCo in WooCommerce-Wear

AliSoftware committed:
Merge pull request #11636 from woocommerce/tooling/wear-app-prototype-build

[Tooling][Wear] Add Prototype Builds for Wear app

iangmaia committed:
Update release lanes to use parameter keys

iangmaia committed:
Remove redundant release_notes.xml (as replaced by `skip_upload_changelogs: false`)

iangmaia committed:
Annotate builds triggered by the current build

iangmaia committed:
Use `fastlane/metadata/android/en-US` as the only source of truth for metadata

iangmaia committed:
Add support for Wear App metadata localization

iangmaia committed:
Delete any other locale changelogs that would still contain the translations for the old release notes

iangmaia committed:
Add comment to conditional skipping of metadata upload

iangmaia committed:
Use `wear_release_note` alone as the key to the Wear App changelog string

iangmaia committed:
Improve code to avoid excess indentation

iangmaia committed:
Fix RuboCop violations

iangmaia committed:
Add comment to `delete_old_changelogs`

iangmaia committed:
Set changelog limit

iangmaia committed:
Preventing errors on CI managing milestones

ThomazFB committed:
Merge branch 'trunk' into feature/introduce-wear-app

iangmaia committed:
Remove files after rebase

ThomazFB committed:
Introduce Refund model

ThomazFB committed:
Introduce initial OrderItem processing inside Refunds

ThomazFB committed:
Declare Wear OrderItem type with dedicated parsing

ThomazFB committed:
Wire OrdersRepository with Refunds Store fetch request

ThomazFB committed:
Add Order Item name mapping to the parsing functions

ThomazFB committed:
Adjust FetchOrderProducts use case to request the Orders Refunds when network is available

ThomazFB committed:
Correctly map non-refunded order Items to obtain ordered products in FetchOrderProducts

ThomazFB committed:
Fetch the Order data before mapping the Ordered Products

ThomazFB committed:
Remove unused code from Refund.kt

ThomazFB committed:
Adjust WearOrder to contain lineItems when orders are sent from the phone

ThomazFB committed:
Introduce Json parsing for LineItems in Order

ThomazFB committed:
Adjust OrderEntityExt.kt missing mapping

ThomazFB committed:
Adjust GetWearableOrderProducts to just get the orders data from cache

ThomazFB committed:
Update FetchOrderProductsTest

ThomazFB committed:
Tweak how line item data is stored and retrieved during the Order Details loading

ThomazFB committed:
Introduce cache loading for Order Refunds

ThomazFB committed:
Overhaul how the FetchOrderProducts retrieve cache data when no network is available

ThomazFB committed:
Add missing emit instruction to FetchOrderProducts data fetching operations

ThomazFB committed:
Fix lint issues

ThomazFB committed:
Fix FetchOrderProductsTest

ThomazFB committed:
Merge pull request #11745 from woocommerce/issue/introduce-better-products-loading

[Watch App] Introduce better products loading

iangmaia committed:
Move PlayStoreStrings.pot to `./fastlane/metadata/`

iangmaia committed:
Added initial wear app release notes

iangmaia committed:
Improve error message when there's an error closing a milestone

iangmaia committed:
Add files' deletion in git when deleting old changelogs

iangmaia committed:
Rename metadata related constants

iangmaia committed:
Fix RuboCop offenses

iangmaia committed:
Use different Github release asset names and version name for the Wear app

iangmaia committed:
Simplify code to delete obsolete changelogs

Co-authored-by: Olivier Halligon <olivier.halligon@automattic.com>

iangmaia committed:
Make sure the deleted files are added to the commit

Co-authored-by: Olivier Halligon <olivier.halligon@automattic.com>

iangmaia committed:
Remove `include_wear_app` parameter from `start_code_freeze`

iangmaia committed:
Merge pull request #11726 from woocommerce/iangmaia/wear-app-localization

[Tooling] Wear App Changelog + Changelog / Metadata improvements

AliSoftware commented:
@ThomazFB  https://github.com/woocommerce/woocommerce-android/pull/11726 has just landed (alongside the accompanying diff in WPCOM to update the MC Release Scenario accordingly—see pdnsEh-1Gh-p2#comment-3678) so I think (at least on our end) everything should now be ready for this PR to be merged before tomorrow's code freeze 😓

ThomazFB committed:
Migrate the entire Wear module package structure

ThomazFB committed:
Fix unit tests imports

ThomazFB committed:
Merge pull request #11760 from woocommerce/issue/adjust-wear-module-package

Issue/adjust wear module package

ThomazFB commented:
Thank you, @AliSoftware, @iangmaia, and @wzieba, for all the work in making this project viable for release! It wouldn't be possible to move forward with the project without all this support. 🙇  I'll be merging this PR in the next hour and making it available in today's code freeze so we can verify how things will behave.
